File: Cuarto/SO/Prácticas/P3/practica3/03-cpu-bound/ult.py
import gevent
import time
from common import nth_prime_number

# Number of threads to spawn
NUM_THREADS = 5

# Delay in seconds to simulate IO-bound work
# adjust for faster results or to have extra time to monitor resources
SLEEP_TIME = 10

# Count the number of primes up to this limit
LIMIT = 500_000


def do_work():
    greenlet_id = id(gevent.getcurrent())
    start = time.time()
    print(f"[{greenlet_id=}] Doing some work...")
    nth_prime_number(LIMIT)
    elapsed = time.time() - start
    print(f"[{greenlet_id=}] Done with work in {elapsed} seconds.")


if __name__ == "__main__":
    start = time.time()
    print("Starting the program.")
    all_greenlets = []
    for i in range(NUM_THREADS):
        # Start NUM_THREADS greenlets
        greenlet = gevent.spawn(do_work)
        all_greenlets.append(greenlet)

    gevent.joinall(all_greenlets)
    elapsed = time.time() - start
    print(f"All greenlets are done in {elapsed} seconds")

File: Cuarto/SO/Prácticas/P3/practica3/03-cpu-bound/ult.c
#include <pth.h>
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include "common.h" // clockwatch_t, difftime_ns, nth_prime_number

#define SLEEP_TIME 10
#define NUM_THREADS 5

// Count the number of primes up to this limit
#define LIMIT 2500000


void* do_work(void* arg) {
    unsigned long thread_id = (unsigned long) pth_self();
    clockwatch_t clock;
    double elapsed;

    
    printf("[Thread %lu] Doing some work...\n", thread_id);
    clockwatch_start(&clock);

    // Calculate the nth prime number
    nth_prime_number(LIMIT); // CPU bound work
    
    elapsed = clockwatch_stop(&clock);
    printf("[Thread %lu] Done with work in %f seconds.\n", thread_id, elapsed);
    return NULL;
}

int main() {
    pth_t threads[NUM_THREADS];
    time_t start, end;
    double elapsed;

    pth_init();

    time(&start);
    printf("Starting the program.\n");

    for (int i = 0; i < NUM_THREADS; i++) {
        threads[i] = pth_spawn(PTH_ATTR_DEFAULT, do_work, NULL);
    }

    for (int i = 0; i < NUM_THREADS; i++) {
        pth_join(threads[i], NULL);
    }

    time(&end);
    elapsed = difftime(end, start);
    printf("All threads are done in %f seconds\n", elapsed);

    pth_kill();

    return 0;
}